### PR TITLE
AVRO-2524 Fix/ignore rubocop error and warnings

### PR DESCRIPTION
**Pull request by mpan-wework**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO/) issues and references them in the PR title. For example, "AVRO-1234: My Avro PR"
  - https://issues.apache.org/jira/browse/AVRO-2524
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason: this is a fix for lint, no change is contained

### Commits

- [x] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

**Comment by Fokko** [comment]
@mpan-wework The CI seems to be broken.

@tjwp Could you do a review on this? My Ruby skills are a bit rusty.

**Comment by nandorKollar** [comment]
The failure is a Rat failure, .rubocop.yml misses license header. @mpan-wework could you please add this new file to Rat excludes [list](https://github.com/apache/avro/blob/master/pom.xml#L286)?

File: lang/ruby/.rubocop.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+Metrics/AbcSize:
+  Enabled: false
+Metrics/CyclomaticComplexity:
+  Enabled: false
+Metrics/PerceivedComplexity:
+  Enabled: false
+Metrics/ParameterLists:
+  Enabled: false
+Naming/UncommunicativeMethodParamName:
+  Enabled: false
```
[the remark this answers is not in the document]
Comment (mpan-wework) [reply lang/ruby/.rubocop.yml]: `Style/Documentation` requires doc in comment describing a class
`Style/NegatedIf` prefers `unless condition` to `if !condition`
`Naming/UncommunicativeMethodParamName` disallows too short variables (e.g. `n`) as method param
`Metrics` stuff restrict like max character length of a line, max line length of a method and a class, max number of method parameters, max numbers of assignments, returns and if-else branches within a method

File: lang/ruby/lib/avro/io.rb [hunk 2]
```diff
@@ -579,21 +579,23 @@ def write_map(writers_schema, datum, encoder)
 
       def write_union(writers_schema, datum, encoder)
         index_of_schema = -1
-        found = writers_schema.schemas.
-          find{|e| index_of_schema += 1; found = Schema.validate(e, datum) }
-        unless found  # Because find_index doesn't exist in 1.8.6
-          raise AvroTypeError.new(writers_schema, datum)
+        found = writers_schema.schemas.find do |e|
+          index_of_schema += 1
+          found = Schema.validate(e, datum)
```
[the remark this answers is not in the document]
Comment (mpan-wework) [reply lang/ruby/lib/avro/io.rb]: removed

File: lang/ruby/lib/avro/io.rb [hunk 3]
```diff
@@ -579,21 +579,23 @@ def write_map(writers_schema, datum, encoder)
...
         end
+        raise AvroTypeError.new(writers_schema, datum) unless found # Because find_index doesn't exist in 1.8.6
```
[the remark this answers is not in the document]
Comment (mpan-wework) [reply lang/ruby/lib/avro/io.rb]: comment removed

File: lang/ruby/lib/avro/io.rb [hunk 4]
```diff
@@ -254,40 +256,39 @@ def read(decoder)
...
-        when :fixed;   read_fixed(writers_schema, readers_schema, decoder)
-        when :enum;    read_enum(writers_schema, readers_schema, decoder)
-        when :array;   read_array(writers_schema, readers_schema, decoder)
-        when :map;     read_map(writers_schema, readers_schema, decoder)
-        when :union;   read_union(writers_schema, readers_schema, decoder)
-        when :record, :error, :request;  read_record(writers_schema, readers_schema, decoder)
-        else
-          raise AvroError, "Cannot read unknown schema type: #{writers_schema.type}"
-        end
+                when :null then    decoder.read_null
```
[the remark this answers is not in the document]
Comment (mpan-wework) [reply lang/ruby/lib/avro/io.rb]: this is a preferred for one-line `when`

**Comment by mpan-wework** [comment]
> Thank you for taking this on! It is definitely a step in the right direction for the Ruby implementation.
> 
> In addition to the comments I've made, there are a couple of other things that should be considered:
> 
> * Does the build install a specific version of rubocop, otherwise builds are likely to get different results depending on the version that is picked up?
> * Currently I think the build runs `rubocop --lint`. If we are attempting to keep the Ruby code consistent with a larger set of cops then they should all be enforced as part of the build.

`rubocop` version to be any recent one would be enough I believe, as far as I have worked on ruby projects. It would implicitly introduce new rule, but it would not change behavior of current rule.

`rubocop` can be categorized into easy-correctable and others. In this PR, I fixed those easy-correctable ones through reorganizing pieces of code. Other ones, as I have mentioned in one of the above comments, they require refactoring techniques like method extraction or use of instance variables.

I would recommend to fix easy-correctable ones first. There are around 1,500 rubocop offenses existing, which could lead to contributors creating new ones unattemptedly. After this round of fixing and ignoring, they can be reduced to 16, then contributors would take care of new code.

File: lang/ruby/lib/avro/logical_types.rb [hunk 5]
```diff
@@ -87,7 +87,7 @@ def self.decode(datum)
     def self.type_adapter(type, logical_type)
       return unless logical_type
 
-      TYPES.fetch(type, {}.freeze).fetch(logical_type, Identity)
```
[the remark this answers is not in the document]
Comment (mpan-wework) [reply lang/ruby/lib/avro/logical_types.rb]: restored

**Pull request closed without merging**